src/timely-util/src/operators_async_ext.rs

```diff
@@ -166,24 +178,32 @@ impl<G: Scope> OperatorBuilderExt<G> for OperatorBuilder<G> {
                 if had_pending_notify && !scheduler.is_notified() {
                     waker.wake_by_ref();
                 }
+
+                operator_incomplete
             }
         });
     }
 }
 
-#[allow(unused_macros)]
+/// Convenience macro used to wrap what might otherwise be the argument to `build_reschedule`.
+#[macro_export]
 macro_rules! async_op {
     (|$capabilities:ident, $frontiers:ident| $body:block) => {
-        move |mut capabilities, mut frontiers, scheduler| async move {
+        move |mut capabilities, frontiers, scheduler| async move {
             loop {
                 scheduler.notified().await;
-                // rebind to mutable references to make sure they can't be accidentally dropped
+
                 #[allow(unused_mut)]
                 let mut $capabilities = &mut capabilities;
-                #[allow(unused_mut)]
-                let mut $frontiers = &mut frontiers;
-                let _: () = async { $body }.await;
+                let $frontiers = (*frontiers.borrow()).clone();
+
+                if !async { $body }.await && $frontiers.iter().all(|f| f.is_empty()) {
+                    break;
+                }
             }
+
+            // Make sure can't be accidentally dropped in `$body`
+            drop(frontiers);
```

Comment:
the `$body` doesn't have a way to affect the `frontiers` binding here due to macro hygiene so we can remove this part

Reply:
ahh yeah true